FirebaseSessions/Sources/Installations+InstallationIDProvider.swift

```diff
@@ -0,0 +1,27 @@
+//
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import FirebaseInstallations
+
+protocol InstallationIDProvider {
+  func provideInstallationID(provide: @escaping (String?, Error?) -> Void)
+}
+
+extension Installations: InstallationIDProvider {
+  func provideInstallationID(provide: @escaping (String?, Error?) -> Void) {
+    installationID(completion: provide)
+  }
+}
```

Comment:
I think we should do three things if we go the protocol route:
1. Hide away the optional completion handler arguments by using `Result` type. It'll make usage easier and more consistent throughout the implementation and tests
2. This one is more personal opinion so feel free to push back: name the protocol `InstallationsProtocol` because it's existence is only to abstract away the `Installations` type for testability's sake. This was a consensus the team arrived at while working on another Swift project. If the protocol is meant to be conformed too by multiple objects, a generic name inline with Swift Protocol's naming guidelines is preferred. If the protocol is mean to abstract away a concrete type for testability then just name the protocol the _concrete type_ + _Protocol_... so here would be _InstallationsProtocol_.

```suggestion
protocol InstallationsProtocol {
  func installationID(completion: @escaping (Result<String, Error>) -> Void)
}

extension Installations: InstallationsProtocol {
    func installationID(completion: @escaping (Result<String, Error>) -> Void) {
        self.installationID { (installationID: String?, error: Error?) in
            if let installationID = installationID {
              completion(.success(installationID))
            } else if let error = error {
              completion(.failure(error))
            }
        }
    }
}

```

Reply:
2 sounds good. My only issue with 1 is it means I can't test that extension - maybe I can just improve the logic in `Identifiers` to use the `if let ...` flow?

Reply:
> My only issue with 1 is it means I can't test that extension

Hmm, I'm not 100% following– there were no visibility changes?

Reply:
Meaning the `InstallationsProtocol` should be visible during testing when doing `@testable import FirebaseSessions`

Reply:
Just pushed up my suggestion

Reply:
Also pushed up the rename

Reply:
Thanks Sam, I still don't follow why we can't use the `func installationID(completion: @escaping (Result<String, Error>) -> Void)` signature?

Reply:
Oh cause I can't test that part of the code

Reply:
Oh there _isn't_ anything to test. We are just wrapping the Installations API with a new API to return the `Result`-type.

Reply:
I guess I don't see how it's any different from the `func provideInstallationID(provide: @escaping (String?, Error?) -> Void) ` except that it has the `Result` parameter as syntactic sugar.

Reply:
Ok updated it